### PR TITLE
[vpj][test] Add minNumberOfSamples to avoid crash in Zstd trainSamples

### PR DESCRIPTION
**Pull request by m-nagarajan**
## Summary
1. Raised an [issue#253](https://github.com/luben/zstd-jni/issues/253) with zstd-jni/zstd which results in the library crashing for some cases with only a few samples. As using just a few samples is not a practical scenario for dict building, added `minNumberOfSamples` to skip creating the dictionary until there is a fix for the issue. As venice is a KV store with a ideally huge data sets, skipping this for a store with very few samples should not affect any stores apart from test stores. This is a failsafe fix as the users can control the sample size in bytes and if that is set to too low, it might lead to this crash. 
2. Added `TestValidateSchemaAndBuildDictMapper` and `Testzstdlibrary`
3. Added more tests in `Testvenicepushjobcheckpoints`

## How was this PR tested?
Github CI

## Does this PR introduce any user-facing changes?
- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/PushJobZstdConfig.java [hunk 1]
```diff
@@ -12,17 +12,26 @@
...
+  private int maxDictSize;
+  private int maxSampleSize;
   private int filledSize; // Duplicate of filledSize in ZstdDictTrainer as there is no getter for this
+  /**
+   * Known <a href="https://github.com/luben/zstd-jni/issues/253">zstd lib issue</a> which
+   * crashes if the input sample is too small. So adding a preventive check to skip training
+   * the dictionary in such cases using minNumberOfSamples. Keeping it simple and hard coding
+   * it as if this check doesn't prevent some edge cases then we can disable the feature itself
+   */
+  private int minNumberOfSamples = 20;
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/PushJobZstdConfig.java:24]: nit: can you make it a private static constant?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/ValidateSchemaAndBuildDictMapper.java [hunk 2]
```diff
@@ -184,23 +184,37 @@ private boolean buildDictionaryAndPersistOutput(
...
+            ByteBuffer compressionDictionary;
+            try {
+              compressionDictionary = ByteBuffer.wrap(inputDataInfoProvider.getZstdDictTrainSamples());
+              mapperOutputRecord.put(KEY_ZSTD_COMPRESSION_DICTIONARY, compressionDictionary);
+              MRJobCounterHelper.incrMapperZstdDictTrainSuccessCount(reporter, 1);
+              LOGGER.info("Zstd compression dictionary size = {} bytes", compressionDictionary.remaining());
+            } catch (Exception e) {
+              MRJobCounterHelper.incrMapperZstdDictTrainFailureCount(reporter, 1);
+              LOGGER.error(
+                  "Training ZStd compression dictionary failed: Maybe the sample size is too small or "
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/ValidateSchemaAndBuildDictMapper.java:212]: This 'Maybe the sample size is too small'  part of the message will be still true given we return early now ?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/PushJobZstdConfig.java (hunk 1 above)
Comment (m-nagarajan) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/PushJobZstdConfig.java]: yes, done.

**Review by m-nagarajan — COMMENTED** [review]
Thanks @majisourav99 for the review. Fixed the comments.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/PushJobZstdConfig.java [hunk 3]
```diff
@@ -12,17 +12,26 @@
...
+  private int maxDictSize;
+  private int maxSampleSize;
   private int filledSize; // Duplicate of filledSize in ZstdDictTrainer as there is no getter for this
+  /**
+   * Known <a href="https://github.com/luben/zstd-jni/issues/253">zstd lib issue</a> which
+   * crashes if the input sample is too small. So adding a preventive check to skip training
+   * the dictionary in such cases using minNumberOfSamples. Keeping it simple and hard coding
+   * it as if this check doesn't prevent some edge cases then we can disable the feature itself
+   */
+  private static int minNumberOfSamples = 20;
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/PushJobZstdConfig.java:24]: make it 'final' and all CAPS to follow the convention.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 4]
```diff
@@ -1876,9 +1877,23 @@ private void updatePushJobDetailsWithMRCounters() {
...
+      LOGGER.info(
+          "pushJobDetails MR Counters: \n\tTotal number of records: {} \n\tSize of keys: {} Bytes "
+              + "\n\tsize of uncompressed value: {} Bytes \n\tConfigured value Compression Strategy: {} "
+              + "\n\tFinal data size stored in Venice (compressed using {}): {} Bytes "
+              + "\n\tData size if compressed using Gzip: {} Bytes \n\tData size if "
+              + "compressed using Zstd with Dictionary: {} Bytes",
+          pushJobDetails.totalNumberOfRecords,
+          pushJobDetails.totalKeyBytes,
+          pushJobDetails.totalRawValueBytes,
+          CompressionStrategy.valueOf(pushJobDetails.valueCompressionStrategy).name(),
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:1892]: Lets skip one, may be the `compressed using` part?

**Review by majisourav99 — DISMISSED** [review]
LGTM

**Pull request merged**